### ISSUE TITLE
Are these duplicate classees okay

### ISSUE THREAD

**Issue by bartkl** [bartkl/cim-to-linkml#22]
```sql
select
	C.Object_ID as ClassID,
	C.Name as ClassName,
	C.Package_ID as PackageID,
	P.Name as PackageName
from t_object as C

inner join t_package as P
on C.Package_ID = P.Package_ID

where Object_Type = "Class"
and Object_ID in <OBJECT_IDS> -- list of object IDs
```

### `IEC61968CIMVersion`
Object IDs 1683, 1760, 2188

ClassID | ClassName | PackageID | PackageName
-- | -- | -- | --
1683 | IEC61968CIMVersion | 100 | DocIEC61968
1760 | IEC61968CIMVersion | 100 | DocIEC61968
2188 | IEC61968CIMVersion | 127 | IEC61968

### `Profile`
Object IDs 846, 2856

ClassID | ClassName | PackageID | PackageName
-- | -- | -- | --
846 | Profile | 54 | GenericDataSet
2856 | Profile | 149 | ExternalInputs

### `ResourceCertification`
Object IDs 2645, 2797

ClassID | ClassName | PackageID | PackageName
-- | -- | -- | --
2645 | ResourceCertification | 139 | InfMarketOperations
2797 | ResourceCertification | 161 | ReferenceData

### `TroubleReportingKind`
Object IDs 1834, 2100

ClassID | ClassName | PackageID | PackageName
-- | -- | -- | --
1834 | TroubleReportingKind | 87 | Customers
2100 | TroubleReportingKind | 87 | Customers

**Comment by bartkl** [comment]
Hi Todd,

I've used this QEA project file: `iec61970cim17v40_iec61968cim13v13b_iec62325cim03v17b_CIM100.1.1.1.qea`

**Comment by bartkl** [comment]
Thanks for the analysis and suggestions!

With regards to the `TroubleReportKind` duplicates, here's the full `*` selection from the `t_object` table:

Object_ID | Object_Type | Diagram_ID | Name | Alias | Author | Version | Note | Package_ID | Stereotype | NType | Complexity | Effort | Style | Backcolor | BorderStyle | BorderWidth | Fontcolor | Bordercolor | CreatedDate | ModifiedDate | Status | Abstract | Tagged | PDATA1 | PDATA2 | PDATA3 | PDATA4 | PDATA5 | Concurrency | Visibility | Persistence | Cardinality | GenType | GenFile | Header1 | Header2 | Phase | Scope | GenOption | GenLinks | Classifier | ea_guid | ParentID | RunState | Classifier_guid | TPos | IsRoot | IsLeaf | IsSpec | IsActive | StateFlags | PackageFlags | Multiplicity | StyleEx | ActionFlags | EventFlags
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1834 | Class | 0 | TroubleReportingKind |   | T. Kostic | 1.0 | Kind of trouble reporting. | 87 | enumeration | 0 | 1 | 0 |   | -1 | 0 | -1 | -1 | -1 | 2008-09-08 13:24:17 | 2022-04-24 12:59:37 | Proposed | 0 | 0 |   | Java |   | 0 |   |   |   |   |   | Java |   |   |   | 1.0 | Public |   |   | 0 | {69FC5C51-2840-4e50-A456-690D9026C079} | 1521 |   |   | 14 | 0 | 0 | 0 | 0 |   |   |   |   |   |  
2100 | Class | 0 | TroubleReportingKind |   | T. Kostic | 1.0 | Kind of trouble reporting. | 87 | enumeration | 0 | 1 | 0 |   | -1 | 0 | -1 | -1 | -1 | 2022-05-14 17:56:44 | 2022-05-14 17:57:34 | Proposed | 0 | 0 |   | Java |   | 0 |   |   |   |   |   | Java |   |   |   | 1.0 | Public |   |   | 0 | {CBCADF0E-A59B-4a75-B085-16E23F8D0229} | 0 |   |   | 14 | 0 | 0 | 0 | 0 |   |   |   |   |   |  

Note that the one with ID `1834` has a parent ID of `1521`. Here's a dump of that object, which turns out to be the `Tariff` class.

Object_ID | Object_Type | Diagram_ID | Name | Alias | Author | Version | Note | Package_ID | Stereotype | NType | Complexity | Effort | Style | Backcolor | BorderStyle | BorderWidth | Fontcolor | Bordercolor | CreatedDate | ModifiedDate | Status | Abstract | Tagged | PDATA1 | PDATA2 | PDATA3 | PDATA4 | PDATA5 | Concurrency | Visibility | Persistence | Cardinality | GenType | GenFile | Header1 | Header2 | Phase | Scope | GenOption | GenLinks | Classifier | ea_guid | ParentID | RunState | Classifier_guid | TPos | IsRoot | IsLeaf | IsSpec | IsActive | StateFlags | PackageFlags | Multiplicity | StyleEx | ActionFlags | EventFlags
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1521 | Class | 0 | Tariff |   |   | 1.0 | Document, approved by the responsible regulatory agency, listing the terms and conditions, including a schedule of prices, under which utility services will be provided. It has a unique number within the state or province. For rate schedules it is frequently allocated by the affiliated Public utilities commission (PUC). | 87 |   | 0 | 1 | 0 |   | -1 | 0 | -1 | -1 | -1 | 2008-05-22 00:21:45 | 2020-03-05 13:58:23 | Proposed | 0 | 0 |   |   |   | 0 |   |   |   |   |   |   |   |   |   |   | Public |   |   | 0 | {0BE74F01-4122-4e79-A801-A70BCEAE82B2} | 0 |   |   | 13 | 0 | 0 | 0 | 0 |   |   |   |   |   |

**Comment by tviegut** [comment]
@bartkl , thanks for narrowing that down.  I caught this exact type of scenario in the past that occurred by accident when a class was accidentally "dragged" into another class (but that wasn't caught at first).  In those situations my guess is that the CIM MM didn't realize it happened or thought it got deleted and ended up creating a second instance - resulting in what we're seeing.  Ironically, EA allows that to occur but I'm not sure why.  I know the EA has a heavy Model Driven Development philosophy in its approach.  Perhaps from that standpoint EA would generate code (e.g. Java, C#, etc.) from the UML that would reflect such an embedded class or enumeration as an "inner class".

**Comment by bartkl** [comment]
Since filtering out using `LIKE Doc%` will also exclude potential packages you don't wish to exclude (e.g. `Document` or something like that). So, safest bet is to hardcode a list of the packages to filter.

**Source QEA**: `CIM_Grid18v10_Support14v00_Market04v14.qea`

## Documentation packages

### Query

```sql
SELECT Package_ID, Name FROM t_package WHERE Name LIKE "Doc%"
```

### Packages to exclude

`Package_ID` | `Name`
-- | --
5 | DocBase
27 | DocDC
189 | DocExtIEC61970
49 | DocICCPConfiguration
70 | DocIEC61968
104 | DocIEC62325

## Informal packages

### Query

```sql
SELECT Package_ID, Name FROM t_package WHERE Name LIKE "Inf%"
```

### Packages to exclude

`Package_ID` | `Name`
-- | --
188 | InfENTSOEextensionsNetworkCodes
62 | InfERPSupport
137 | InfExternalInputs
224 | InfExtProtection
50 | InfFeeder
143 | InfFinancial
30 | InfGenericDataset
31 | InfGenericDatasetInstanceDiagrams
32 | InfGenericDatasetInstanceExamples
29 | InfGrid
160 | InfHVDCDynamics
75 | InfIEC61968
132 | InfIEC62325
64 | InfLocations
133 | InfMarketOpCommon
138 | InfMarketOperations
139 | InfMarketResults
99 | InfMetering
100 | InfNewAssets
92 | InfOperations
54 | InfPart303
140 | InfParticipantInterfaces
195 | InfProtectionControlExtentions
134 | InfReferenceData
144 | InfReservation
59 | InfTypeAsset
88 | InfWiresExt
60 | InfWork

**Comment by bartkl** [comment]
Update: the CIM Modeling Guidelines actually state:

> “Inf” is the prefix used to denote informative CIM UML sub-packages. This is to avoid name clashes among normative and informative packages (e.g., ‘Core’ vs. ‘InfCore’), and to make informative sub-packages obvious through their name. Note that this will apply to all packages that start with this prefix (e.g., 'InfCore' as well as 'Informative').

So I am allowed to make the assumption and work with `NOT LIKE "Inf%"`.

**Comment by bartkl** [comment]
Issue: some subpackages of doc/inf packages are not themselves doc or inf. These also need to be excluded however.

**Comment by bartkl** [comment]
Done. Last remaining check: exclude private/internal packages. Here too holds, I think, that private ancestors imply you are private as well? Check with Todd.

**Comment by ucaiug-admin** [comment]
> Done. Last remaining check: exclude private/internal packages. Here too holds, I think, that private ancestors imply you are private as well? Check with Todd.

@bartkl :  You are right, child packages not prefixed with Inf* (though they should be) should be treated as Inf packages.  May require special query to get this correct I am guessing

**Comment by ucaiug-admin** [comment]
> Done. Last remaining check: exclude private/internal packages. Here too holds, I think, that private ancestors imply you are private as well? Check with Todd.

@bartkl :  You are right, child packages not prefixed with Inf* (though they should be) should be treated as Inf packages.  May require special query to get this correct I am guessing
